chirp/drivers/retevis_ha1g.py

```diff
@@ -0,0 +1,3105 @@
+
+
+def handshake(self, serial):
+    databytes = b""
+    num = 0
+    while num <= 5:
+        serial.write(get_handshake_bytes(self.MODEL + " "))
+        databytes = serial.read(self.page_len)
+        flag, databytes = handle_connect_ver(
+            databytes, serial
+            )
+        if flag:
+            break
+        num += 1
+    return handshake_handle_connect_event(self, databytes)
+
+
+def read_items(self, serial):
+    all_bytes = bytearray(self._memsize)
+    status = chirp_common.Status()
+    status.msg = "Cloning from radio"
+    status.cur = 0
+    status.max = self._memsize
+    for item in Clone_TypeEnum:
+        try:
+            item_bytes = get_read_current_page_bytes(
+                self, item.value, serial, status
+                )
+            if item_bytes:
+                all_bytes = read_item_handle_connect_event(
+                    all_bytes, item_bytes, item
+                    )
+        except Exception as e:
+            LOG.error(
+                f"read item_data error: {item.name} error_msg: {e}"
+                )
+            continue
+    return all_bytes
+
+
+def write_items(self, serial):
+    status = chirp_common.Status()
+    status.msg = "Uploading to radio"
+    status.cur = 0
+    status.max = self._memsize
+    data_bytes = self.get_mmap()
+    for item in Clone_TypeEnum:
+        if (
+            item == Clone_TypeEnum.radioHead
+            or item == Clone_TypeEnum.radioInfo
+            or item == Clone_TypeEnum.radioVer
+            or item == Clone_TypeEnum.zoneData
+            or item == Clone_TypeEnum.outFactoryData
+        ):
+            continue
+        item_bytes = get_write_item_bytes(data_bytes, item)
+        write_item_current_page_bytes(
+            self, serial, item_bytes, item.value, status
+            )
+
+
+def handshake_handle_connect_event(self, dataByte: bytes):
+    if dataByte[14:16] == b"\x00\x01":
+        if dataByte[20] != 1:
+            model_str = (
+                self.MODEL.decode("ascii")
+                if isinstance(self.MODEL, bytes)
+                else str(self.MODEL)
+            )
+            radioType = radioType = dataByte[20: 20 + len(model_str)]
+            if radioType == f"{model_str}".encode("ascii"):
+                return HandShakeStuts.Normal
+            else:
+                return HandShakeStuts.RadioWrong
+        else:
+            return HandShakeStuts.PwdWrong
+    else:
+        return HandShakeStuts.Wrong
+
+
+def handle_connect_ver(
+    current_page_byte: bytes, serial_conn, max_retries=5, chunk_size=1024
+) -> tuple[bool, bytes]:
+    new_bytes = current_page_byte
+    if not new_bytes:
+        return False, new_bytes
+    for _ in range(max_retries):
+        if rdtp_page_data_crc16_ver(new_bytes):
+            return True, new_bytes
+        chunk = serial_conn.read(chunk_size)
+        if not chunk:
+            return False, new_bytes
+        new_bytes += chunk
+    return False, new_bytes
```

Comment:
What this this function doing? It looks like maybe it's reading data until the last byte ends in a valid crc for the early bytes? Please don't do stuff like this. Read the amount of data you're expecting, then calculate the checksum with the expected amount. This sort of algorithm lends itself to non-deterministic behavior, especially across the multiple platforms that CHIRP supports.

Reply:
It looks to me like this (and the inner) function does this:

1. Sends some data
2. Reads `chunk_size` data
3. Guesses if the result should be CRC'd based on the length and whether or not the CRC matches
4. Returns the data it read and True/False based on what? Whether the CRC matched or not?

This is not a good pattern. You should send the thing you want, and read what you expect. If you expect the thing returned to be CRC'd then do it, don't make this function guess based on the length of the thing read.

Looks like `max_retries` is not used? Please remove.